pkg/agent/multicast/mcast_controller_test.go

```diff
@@ -621,6 +622,115 @@ func TestNodeUpdate(t *testing.T) {
+						OFPort: int32(i),
+					},
+				},
+			}
+		}
+	}()
+	// Below func adds local group leave events.
+	go func() {
+		defer wg.Done()
+		LeastSignificantByteArr := rand.Perm(numEvents)
+		for i := 0; i < len(LeastSignificantByteArr); i++ {
+			c.groupEventCh <- &mcastGroupEvent{
+				group: groupIP,
+				eType: groupLeave,
+				time:  time.Now(),
+				iface: &interfacestore.InterfaceConfig{
+					Type:          interfacestore.ContainerInterface,
+					InterfaceName: fmt.Sprintf("local-interfaceName-%d", i),
+					OVSPortConfig: &interfacestore.OVSPortConfig{
+						OFPort: int32(i),
+					},
+				},
+			}
+		}
+	}()
+	// Below func adds remote group join events.
+	go func() {
+		defer wg.Done()
+		LeastSignificantByteArr := rand.Perm(numEvents)
+		for i := 0; i < len(LeastSignificantByteArr); i++ {
+			c.groupEventCh <- &mcastGroupEvent{
+				group: groupIP,
+				eType: groupJoin,
+				time:  time.Now(),
+				iface: &interfacestore.InterfaceConfig{
+					Type:          interfacestore.TunnelInterface,
+					InterfaceName: fmt.Sprintf("remote-interfaceName-%d", i),
+					OVSPortConfig: &interfacestore.OVSPortConfig{
+						OFPort: int32(i),
+					},
+				},
+				srcNode: net.ParseIP(fmt.Sprintf("10.20.30.%d", LeastSignificantByteArr[i]+2)),
+			}
+		}
+	}()
+	// Below func adds remote group leave events.
+	go func() {
+		defer wg.Done()
+		LeastSignificantByteArr := rand.Perm(numEvents)
+		for i := 0; i < len(LeastSignificantByteArr); i++ {
+			c.groupEventCh <- &mcastGroupEvent{
+				group: groupIP,
+				eType: groupLeave,
+				time:  time.Now(),
+				iface: &interfacestore.InterfaceConfig{
+					Type:          interfacestore.TunnelInterface,
+					InterfaceName: fmt.Sprintf("remote-interfaceName-%d", i),
+					OVSPortConfig: &interfacestore.OVSPortConfig{
+						OFPort: int32(i),
+					},
+				},
+				srcNode: net.ParseIP(fmt.Sprintf("10.20.30.%d", LeastSignificantByteArr[i]+2)),
+			}
+		}
+	}()
+	for i := 0; i < numEvents; i++ {
+		mockIfaceStore.EXPECT().GetInterfaceByName(fmt.Sprintf("local-interfaceName-%d", i)).Return(&interfacestore.InterfaceConfig{
+			Type:          interfacestore.ContainerInterface,
+			InterfaceName: fmt.Sprintf("local-interfaceName-%d", i),
+			OVSPortConfig: &interfacestore.OVSPortConfig{
+				OFPort: int32(i),
+			},
+		}, true).AnyTimes()
+	}
+
+	mockOFClient.EXPECT().InstallMulticastGroup(gomock.Any(), gomock.Any(), gomock.Any()).AnyTimes()
+	mockOFClient.EXPECT().InstallMulticastFlows(groupIP, gomock.Any()).AnyTimes()
+	mockOFClient.EXPECT().UninstallMulticastGroup(gomock.Any()).AnyTimes()
+	mockOFClient.EXPECT().UninstallMulticastFlows(groupIP).AnyTimes()
+	mockOFClient.EXPECT().SendIGMPRemoteReportPacketOut(gomock.Any(), gomock.Any(), gomock.Any()).AnyTimes()
+	mockOFClient.EXPECT().SendIGMPQueryPacketOut(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).AnyTimes()
+	go c.eventHandler(stopCh)
+	for i := 0; i < 2; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	wg.Wait()
```

Comment:
there is no guaratee that when this call finishes, eventHander and worker have processed the events. Could we use some determinstic input and decide when the goroutines finish their job by polling the result, which could be more determinstic?

Reply:
Added a check on the size of eventCh and queue after all events have been sent.